### DISCUSSION TITLE
RPi Pico supported?

### DISCUSSION THREAD

**Discussion by GitFun4All** [bxparks/AceButton#80] [Q&A] [answered]
Hi, does anyone use AceButton on the Pico (Arm M0+ based)?

**Comment by bxparks** [comment, marked as answer]
I don't have a RPi Pico so I have not tested AceButton on it.

As far as I can tell, there are 2 Arduino cores for the RP2040:

* https://github.com/arduino/ArduinoCore-mbed
* https://github.com/earlephilhower/arduino-pico

and both of them are based on https://github.com/arduino/ArduinoCore-api, which has numerous backwards incompatibility issues with the previous Arduino API. So AceButton may work accidentally for the RPi Pico, but I don't plan on supporting the RPi officially.

**Reply by GitFun4All** [reply]
Hi - thank you for the detailed answer. I’ll install the Arduino firmware (they come with micro python) and try it.

I’ll happily post you one if it would tempt you into supporting it..... ;-)

**Reply by bxparks** [reply]
My guess is that AceButton would work, because I had to make some changes to support the Nano Every, which uses new ArduinoCore-api, in which they broke backwards compatibility of `digitalRead()` (!).  But I don't have a Nano Every, so someone else had to do the debugging and testing.

I would like to support as many hardware platforms as possible, and the RPi Pico looks really interesting. But the backwards compatibility breakages of the ArduinoCore-api are substantial, surprising, and undocumented. I'd rather not spend my limited bandwidth chasing the unknown compatibility issues of the new Arduino API, on hardware that I don't own, and don't use. I have a hard enough time just keeping up with the changes in the Arduino Cores of the hardware that I do support. For example, the STM32duino recently upgraded from 1.9.0 to 2.0.0, and the ESP8266 just upgraded from 2.7.4 to 3.0.0. I need to go through all my libraries and re-validate them on those new versions. But I haven't had the time...

**Reply by GitFun4All** [reply]
Just a quick update to confirm Acebutton version 1.9.0 does seem to work well with the Pi-pico and the ArduinoCore-mbed 2.0 core as far as I can tell. 
I've not done any extensive testing, but the library examples for the simple demo button definitely works as predicted and so far the GPIO + internal LED mapping seems sane. 
 
Thanks again for developing & sharing Acebutton, I still mostly love it! ;D
